### DISCUSSION TITLE
Not loading states properly

### DISCUSSION THREAD

**Discussion by MohammadZarifiyan** [spatie/laravel-model-states#246] [Q&A]
Hello
What is this code problem?

spatie/laravel-model-states: ^2.6
laravel/framework: ^9.24

```php
 App\Models\WithdrawRequest::getStates()
// Throws Spatie\ModelStates\Exceptions\InvalidConfig  No states registered for Spatie\ModelStates\State.
```
```php
<?php

namespace App\Models;

use App\State\WithdrawRequest\WithdrawRequestStatus;
use Illuminate\Database\Eloquent\Model;
use Spatie\ModelStates\HasStates;

class WithdrawRequest extends Model
{
    use HasStates;

    protected $fillable = ['status'];

    protected $casts = [
        'status' => WithdrawRequestStatus::class,
    ];
}

```
```php
<?php

namespace App\State\WithdrawRequest;

use App\Events\WithdrawRequestStatusChanged;
use Spatie\ModelStates\State;
use Spatie\ModelStates\StateConfig;

class WithdrawRequestStatus extends State
{
    public static function config(): StateConfig
    {
        return parent::config()
            ->allowAllTransitions()
            ->stateChangedEvent(WithdrawRequestStatusChanged::class);
    }
}
```
```php
<?php

namespace App\State\WithdrawRequest;

class Pending extends WithdrawRequestStatus
{
    public static string $name = 'pending';
}
```
```php
<?php

namespace App\State\WithdrawRequest;

class Paid extends WithdrawRequestStatus
{
    public static string $name = 'paid';
}
```
```php
<?php

namespace App\State\WithdrawRequest;

class Rejected extends WithdrawRequestStatus
{
    public static string $name = 'rejected';
}
```
Also I've used `->registerState()` in WithdrawRequestStatus.php, Error disappears but `getStatesFor('status')` returns empty collection. But `WithdrawRequestStatus::all()` works properly and returns full states list (`Pending`/`Rejected`/`Paid`)

**Comment by perisicnikola37** [comment]
Hi there,

Which problem are you having with this code?

**Reply by jonjakoblich** [reply]
The issue he is having is the same I came here to research. The package automatically registers states located in the same directory as the abstract state, however in order to use `allowAllTransitions()` in the config method, you have to explicitly register states. This seems like a flaw in the implementation of `allowAllTransitions(`) to me, and takes away from the beauty of having all your states automatically registered.

I may open a PR to fix this, but I need to dig into the code a little more. I can see how `allowAllTransitions()` was implemented in the original PR #238, but it feels flawed with having to explicitly register states. You are torn between explicitly registering all the states or explicitly registering all the transitions.

As I type (and think) through this, one possibility might be to implement a `registerAllStates(string $directory = __DIR__)` method where you optionally pass in the directory where all the states are located and the method will do the work of all the registrations. Calling this in combination with `registerAllTransitions()` will remove some of the burden. There are some initial threads of this that can be gleaned from #156. 

Plenty to think about.

**Comment by jonjakoblich** [comment]
Created a PR to address this issue #265

**Comment by Maxiviper117** [comment]
I am getting this same issue when trying to use `allowAllTransitions()`, with  similar setup.



I get this error

```bash
   Spatie\ModelStates\Exceptions\InvalidConfig 

  No states registered for App\States\Subscription\SubscriptionState

  at vendor\spatie\laravel-model-states\src\StateConfig.php:136
    132▕      */
    133▕     public function allowAllTransitions(): StateConfig
    134▕     {
    135▕         if (empty($this->registeredStates)) {
  ➜ 136▕             throw new InvalidConfig('No states registered for ' . $this->baseStateClass);
    137▕         }
    138▕
    139▕         $this->allowTransitions(collect($this->registeredStates)->crossJoin($this->registeredStates)->toArray());
    140▕

  1   app\States\Subscription\SubscriptionState.php:17
      Spatie\ModelStates\StateConfig::allowAllTransitions()

  2   vendor\spatie\laravel-model-states\src\HasStates.php:121
      App\States\Subscription\SubscriptionState::config()
 ```
      
Until #265 is merged, what is the way to allow all states to transition to any other state?

**Reply by jonjakoblich** [reply]
You have to explicitly register states prior to using `allowAllTransitions()`

**Comment by jonjakoblich** [comment]
Great news! Looks like my PR #265 was approved and merged into the package. This is no longer an issue.
